Fix some 3d jitter issues by changing contact caching and friction clamping

Geekotron opened the pull request:
fix #61827
fix #62819

Fixes to the 3d physics body pair solver to eliminate jitter. Cause of the jitter was found to be feedback between normal impulse & friction impulse causing the friction clamping to infinitely apply the same impulse in opposite directions each solver iteration. Also, the existing approach to contact caching/warm starting was causing some unnecessary extra impulses/iterations. 
Changes
* split friction into 2 clamped lambda amounts instead of a single vector
* reproject the friction tangents each frame so changes to normal direction would be reflected in accumulated lambda
* no longer leave outdated contacts in the body pair's contact list. they are either used to warm start a current contact or discarded
* perform tangent impulses in a 2nd contact loop after all normal impulses to prevent friction from causing an imbalance in normal impulses

~~Even though the soft body solver uses a very similar approach, I did _not_ update that solver since I haven't seen any issues reported with it. If anyone feels we _should_ alter soft body to have the same changes, let me know and I'll add them there, too.~~

Disclaimer - I had very little experience with these algorithms before a couple weeks ago, and relied heavily on information posted by Erin Catto & Dirk Gregorius. As well as looking at examples in other open source projects such as Jolt Physics.

Geekotron commented:
Test results 

https://user-images.githubusercontent.com/1498979/210184781-4a8221a3-b60c-4058-ad20-218221dbef80.mp4

https://user-images.githubusercontent.com/1498979/210184785-ae9d862e-df94-4c87-8d84-034146975e2c.mp4

https://user-images.githubusercontent.com/1498979/210184789-0d19f9ae-62e3-4a03-b0f2-aa71f914636f.mp4

I also confirmed the rigidbody3d unit test from https://github.com/fabriceci/Godot-Physics-Tests , I did need to increase the test duration on box & poly a few seconds, but I think it's an improvement because a) they go to sleep and I think the results look better; the stack is more vertical after settling down instead of leaning like the old one did and b) I can increase the stack size to 10 where it would previously collapse and it does not collapse after this change.

AThousandShips commented:
You need to run clang-format on your source, fixing each issue as it arises really draws a lot of resources from the CI

fire commented:
I think if the changes are the same for soft body they should be synced.

AThousandShips commented on servers/physics_3d/godot_body_pair_3d.cpp:
```diff
@@ -251,7 +239,10 @@ bool GodotBodyPair3D::setup(real_t p_step) {
 
-	validate_contacts();
+	// reset contacts, keep prior frame's contacts as cache for warm start.
+	std::swap(prior_contacts, contacts);
```
I believe the engine provided `SWAP` is preferred over the standard library one.

Geekotron replied:
SWAP doesn't seem to support arrays. Compiler errors occur if I change it to SWAP:
`
typedefs.h(135): error C3074: an array cannot be initialized with a parenthesized initializer
typedefs.h(136): error C3863: array type 'GodotBodyContact3D::Contact [4]' is not assignable
`

Geekotron commented:
@AThousandShips , sorry about that - my Visual Studio doesn't seem to be respecting the clang-format file. I'll just run it from command line from now on. Also added a whitespace remover extension to handle the checks in file-format.sh

@fire , I'll take a look at soft body. But first I'll need to set up some test cases to ensure it behaves the same after changes, unless you know of an existing project containing any soft body regression tests... ?

Geekotron committed:
Fix some 3d jitter issues by changing contact caching and friction clamping

 (rebased to incorporate impulse reporting from 8ee6264cc914fe1d007faf440fe4fda28b567cc6)

Geekotron commented:
I've rebased this to eliminate conflicts (i.e. to incorporate Lily's impulse reporting changes).

Rudolph-B commented:
@Geekotron Hi, can I update this to remove the merge conflicts/get it in a state to be pulled?

Geekotron commented:
> @Geekotron Hi, can I update this to remove the merge conflicts/get it in a state to be pulled?

I sadly don't have time to continue work on this PR currently, you are certainly welcome to take it over.

AThousandShips reviewed (commented):
For anyone taking this over please consider these
